lighthouse-core/report/html/renderer/report-ui-features.js

```diff
@@ -391,8 +408,67 @@ class ReportUIFeatures {
    */
   onToolsButtonClick(e) {
     e.preventDefault();
-    this.toolsButton.classList.toggle('active');
-    this._document.addEventListener('keydown', this.onKeyDown);
+    e.stopImmediatePropagation();
+
+    if (this.toolsButton.classList.contains('active')) {
+      this.closeToolsDropdown();
+    } else {
+      this.openToolsDropdown();
+    }
+  }
+
+  /**
+   * Handler for tool button.
+   * @param {KeyboardEvent} e
+   */
+  onToolsButtonKeydown(e) {
+    switch (e.code) {
+      case 'ArrowUp':
+      case 'ArrowDown':
+      case 'Enter':
+      case ' ':
+        e.preventDefault();
+        this.openToolsDropdown();;
+        break;
+      default:
+       // no op
+    }
+  }
+
+  /**
+   * Handler for tool dropdown.
+   * @param {KeyboardEvent} e
+   */
+  onToolsDropDownKeydown(e) {
+    const el = /** @type {?Element} */ (e.target);
+
+    if (!el) {
+      return;
+    }
+
+    let nextEl = el.nextElementSibling || this._dom.find(':first-child', this.toolsDropDown);
+    let prevEl = el.previousElementSibling || this._dom.find(':last-child', this.toolsDropDown);
+
+    if ((nextEl.getAttribute('disabled') !== null) || (window.getComputedStyle(nextEl).display === 'none')) {
+      nextEl = nextEl.nextElementSibling;
+    }
+
+    if ((nextEl.getAttribute('disabled') !== null) || (window.getComputedStyle(prevEl).display === 'none')) {
+      prevEl = prevEl.previousElementSibling;
+    }
```

Comment:
This is my least favorite code. it really relies on the DOM structure, checking for disabled/display="none" is for the "Save as gift" option which gets disabled or hidden depending on the host of the report and if you have saved before.

I think a better approach would be to manage the state in JS instead of in the DOM but that would require a larger refactor (see my comment on line 401)

Reply:
imo I think it's fine to use `first-child`/`last-child` and certainly prefer it to maintaining state that the DOM already has.